### DISCUSSION TITLE
Problem with s3 region solved in k8s and third party block storage (S3), but fails with Fusion

### DISCUSSION THREAD

**Discussion by dariogf** [nextflow-io/nextflow#4469] [General]
Hi all,

I would like to know your experience before opening an issue, just in case I am having some mistake.

I am totally new to nextflow, but have made progress, I already have a local, slurm and Google Batch executors working fine, and now I am trying to use it with a managed k8s from OVH, and using their own custom block storage (with S3 interface), not an AWS bucket. 

That means that I have to provide a custom endpoint and also a custom region for S3 to work. 

In my config file, I have:

```
profiles {


    ovh {
        workDir = 's3://mybucket/testwf/work'
        
        docker.enabled = true
        wave.enabled = true
        fusion.enabled = true
        fusion.exportStorageCredentials = true

        aws {
        
            accessKey = 'XXXXXXXXXXXXXXXXXXXXXX'
            secretKey = 'YYYYYYYYYYYYYYYYYYYYYY'
            region = 'gra'
            
            client {
                endpoint = 'https://s3.gra.io.cloud.ovh.net'
                s3PathStyleAccess = true
                protocol = 'https'
            }
        
        }

        k8s {
            namespace = 'default'
            computeResourceType = 'Job'
        }

        process {
            executor = 'k8s'
            scratch = false
            ...
        }
    }
}
```

With this configuration, I get this error, that means that nextflow is not honoring my configuration with 'gra' region:

```
ERROR ~ The authorization header is malformed; the region 'us-east-1' is wrong; expecting 'gra' (Service: Amazon S3; Status Code: 400; Error Code: AuthorizationHeaderMalformed; Request ID: tx148dbb3399004be684136-006542a3b1; S3 Extended Request ID: null; Proxy: null)
```

The same error raises when I remove the aws configuration and credentials from the config file and put them on `.aws/credentials`. Command line aws works fine, so credentials are ok.

I have forked nextflow and looking at the code, I think that I have narrowed the origin of this problem to this line in 
[S3FileSystemProvider](https://github.com/nextflow-io/nextflow/blob/fe5bea992e2c0dd99a2f72388df386975c32c83a/plugins/nf-amazon/src/main/nextflow/cloud/aws/nio/S3FileSystemProvider.java#L839)

```
        final String bucketName = S3Path.bucketName(uri);
        final boolean global = bucketName!=null;
        final AwsClientFactory factory = new AwsClientFactory(awsConfig, Regions.US_EAST_1.getName());
        client = new S3Client(factory.getS3Client(clientConfig, global));

```

The `new AwsClientFactory` line pass a fixed region (US_EAST_1) as parameter, and inside the [constructor](https://github.com/nextflow-io/nextflow/blob/fe5bea992e2c0dd99a2f72388df386975c32c83a/plugins/nf-amazon/src/main/nextflow/cloud/aws/AwsClientFactory.groovy#L112) this fixed region is assigned in first place to the s3client region:

```
        this.region
                = region  // this is the Regions.US_EAST_1.getName() parameter
                ?: config.region
                ?: SysEnv.get('AWS_REGION')
                ?: SysEnv.get('AWS_DEFAULT_REGION')
                ?: fetchRegion()
```

I would expect that the assignment of the fixed region parameter to be at the end of the assignment block, and the first one to be the one from the config, so I have modified it like this:

```
diff --git a/plugins/nf-amazon/src/main/nextflow/cloud/aws/AwsClientFactory.groovy b/plugins/nf-amazon/src/main/nextflow/cloud/aws/AwsClientFactory.groovy
index b60c1d794..78a259645 100644
--- a/plugins/nf-amazon/src/main/nextflow/cloud/aws/AwsClientFactory.groovy
+++ b/plugins/nf-amazon/src/main/nextflow/cloud/aws/AwsClientFactory.groovy
@@ -109,11 +109,11 @@ class AwsClientFactory {
 
         // -- get the aws default region
         this.region
-                = region
-                ?: config.region
+                = config.region
                 ?: SysEnv.get('AWS_REGION')
                 ?: SysEnv.get('AWS_DEFAULT_REGION')
                 ?: fetchRegion()
+                ?: region
 
         if( !this.region )
             throw new AbortOperationException('Missing AWS region -- Make sure to define in your system environment the variable `AWS_DEFAULT_REGION`')
```

After `make` and `make pack`, the staging of files works as expected with the fixed build, but a new problem arrises with Fusion.

Pods and jobs are correctly created on the k8s cluster, but the Fusion file system fails with an error saying that region is empty and that it is not 'gra'. So I suspect that the fusion binary is not taking into account the region configuration, as it happens with nextflow. 

I have also tried to define `AWS_REGION`, and `AWS_DEFAULT_REGION` manually on the pod env, and the variables are populated to the pod, but they are not honored by the fusion binary:

```
10:15AM WRN retrying after error at 'singleStore.ListDirectory error="operation error S3: ListObjectsV2, https response error StatusCode: 400, RequestID: tx1b5aea4b083f484b82d78-0065422534, HostID: tx1b5aea4b083f484b82d78-0065422534, api error AuthorizationHeaderMalformed: The authorization header is malformed; the region '' is wrong; expecting 'gra'" RetryOnError=0
```

Please, someone with access to the fusion source code could see if region can be set? 

Someone know a workaround of how can I tell fusion binary the region to use?

**Comment by pditommaso** [comment]
Excellent troubleshooting. Regarding the region in the [AwsClientFactory](https://github.com/nextflow-io/nextflow/blob/159effb1cf6dfee98ef8647d9bee1ce6df15fcfd/plugins/nf-amazon/src/main/nextflow/cloud/aws/nio/S3FileSystemProvider.java#L839-L839), the rationale is to force the us-west-1 region for "global" S3 operations. However, I agree that makes no sense in your case. You want to open an issue for it. 

In the case of region propagation for Fusion, I think you are right. Looking at [this code](https://github.com/nextflow-io/nextflow/blob/159effb1cf6dfee98ef8647d9bee1ce6df15fcfd/plugins/nf-amazon/src/main/nextflow/cloud/aws/fusion/AwsFusionEnv.groovy#L39-L52) it seems the AWS_REGION is not added to the Fusion environment. @jordeu What do you think?

**Reply by dariogf** [reply]
Thanks, will open an issue.

**Comment by jordeu** [comment]
`AWS_REGION` is not use by Fusion when you provide a custom endpoint. Double check that `AWS_S3_ENDPOINT` is define at your pod.

All the providers that I've tested didn't require the region to be set (the endpoint already fixes the region). I'm going to prepare a snapshot version using the `AWS_REGION` if present, so you can test if it work on OVH.

**Comment by jordeu** [comment]
Give it a try to this Fusion snapshot running Nextflow with:

```
export FUSION_CONTAINER_CONFIG_URL=https://fusion-develop.s3.eu-west-2.amazonaws.com/snapshots/2023/11/02/060449964/fusion-snapshot-amd64.json
```

Double check that `AWS_S3_ENDPOINT` and `AWS_REGION` both are set at the pod.

**Reply by dariogf** [reply]
Thanks you both, I am impressed by your fast responses!

I tried it with the same result:

```
8:16AM WRN retrying after error at 'singleStore.ListDirectory error="operation error S3: ListObjectsV2, https response error StatusCode: 400, RequestID: tx68546741d17949f7a9af3-0065435ac9, HostID: tx68546741d17949f7a9af3-0065435ac9, api error AuthorizationHeaderMalformed: The authorization header is malformed; the region '' is wrong; expecting 'gra'" RetryOnError=0
```

To be sure that I was using the correct binary on the pod, I have checked that the fusion binary used inside the pod is the same than the one pointed by `FUSION_CONTAINER_CONFIG_URL` :

```
pod# md5sum fusion
1e6a1fdaf8c440cae3fe2d1a7144d808  fusion
```


AWS environment into the pod:

```
# printenv | grep -i aws
AWS_S3_ENDPOINT=https://s3.gra.io.cloud.ovh.net
AWS_REGION=gra
AWS_SECRET_ACCESS_KEY=XXXXXXXXXXXXXX
AWS_ACCESS_KEY_ID=YYYYYYYYYYYYYYY
```

Just for completeness, I have also forced AWS_DEFAULT_REGION and extra env variables within pod definition in config file, with the same results:

```
process {
      executor = 'k8s'

      pod = [
          [env: 'AWS_DEFAULT_REGION', value: 'gra'],
          [env: 'AWS_REGION', value: 'gra'],
          [env: 'AWS_S3_REGION', value: 'gra']
      ]
      scratch = false
    }
```

With this the environment in the pod is:

```
pod# printenv | grep -i aws
AWS_S3_ENDPOINT=https://s3.gra.io.cloud.ovh.net
AWS_DEFAULT_REGION=gra
AWS_REGION=gra
AWS_SECRET_ACCESS_KEY=XXXXXXXXXXXXXX
AWS_S3_REGION=gra
AWS_ACCESS_KEY_ID=YYYYYYYYYYYYYYY
```

But raises the same error than before. 

Feel free to ask me for any test.

**Reply by jordeu** [reply]
Ok, I think there was one missing thing on the previous snapshot. Please give it a try to this one:

```
export FUSION_CONTAINER_CONFIG_URL=https://fusion-develop.s3.eu-west-2.amazonaws.com/snapshots/2023/11/02/170440846/fusion-snapshot-amd64.json
```

**Reply by dariogf** [reply]
It seems to be mounting it OK now! 

Tomorrow I will launch a complete workflow, and share my results.

Thanks.

**Reply by dariogf** [reply]
Hi there!,

I have launched a complete workflow and it has worked as expected. Nextflow now works with managed Kubernetes clusters and S3 storage from vendors that needs custom regions to be used:

```
Completed at: 03-Nov-2023 10:36:21
Duration    : 2h 6m 35s
CPU hours   : 3.8
Succeeded   : 28
```

I suppose that those changes will be incorporated at some point into the main branch... meanwhile I will keep using my modified nextflow version and the fusion snapshot that you have provided.

Thanks for your help.

**Reply by jordeu** [reply]
At the end of next week we are going to release a new version with this change. 

We may use `FUSION_AWS_REGION` instead of `AWS_REGION` as variable to set the Fusion region, but this should be transparent because is Nextflow who is going to set these variables.

**Reply by dariogf** [reply]
That`s great !

**Reply by jordeu** [reply]
@dariogf Fusion v2.2.8 with this fix has been released and is automatically used by Nextflow unless you fix it using `export FUSION_CONTAINER_CONFIG_URL=...` to a different version.

Notice that the final variable used is `FUSION_AWS_REGION` and not `AWS_REGION` and you need to manually add it until next Nextflow release is available that will automatically add it:
https://github.com/nextflow-io/nextflow/commit/8f8b09fa40ed7ef40d2f64ea70e7f695ebb31754

**Reply by dariogf** [reply]
Thanks, have just tried it and I can confirm that it is working as expected (with my own modified version of nextflow).
